### PR TITLE
Global Styles: Try a simple navigation

### PR DESCRIPTION
**Pull request by tyxla**
## Description

One of the bigger blockers for moving #32923 forward (moving global styles v2 from g2 to Gutenberg) is the navigation/router logic. There's a lot to discuss there when it comes to how functional and "routery" we want it to be, whether we need advanced routing (with react router or so), URL persistence, history management, a dedicated data store, animations, transitions, etc. However, to me, none of that is a blocker for the initial implementation and all of those could be incrementally added to the project. 

So what we really need to move the project forward is a simple navigation system that's responsible for holding all the views (I'm adopting that term [as per this discussion](https://github.com/WordPress/gutenberg/pull/32923#discussion_r657084553)), maintaining and presenting the current view and allowing navigation between all the views. That's all we need in order to make progress for now, and all the rest could be considered a feature and/or eye candy. If we land this, it will help unblock much of #32923 and allow us to actually get much closer to a working prototype.

So this PR sets the stage with a very simple navigation system that's based on the existing `Navigation` component base and uses the Context API. We're introducing the views that are introduced in #32923 as simple empty placeholder components with some navigational items inside them, which we're currently using just for demonstration. Since that's not imported anywhere, we can use that stage and further build the Global Styles v2 prototype on top of it.

## How has this been tested?
This is manually testable through a new Storybook story:

* Checkout this branch
* Run `npm run storybook:dev`
* Navigate to the "Edit Site (experimental)" > "Global Styles v2" > "Default" story.
* Play with the navigation buttons and verify you can navigate as expected.

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+import { VIEW_ROOT } from './constants';
+import { GlobalStylesNavigationContext } from './context';
+
+export default function GlobalStylesNavigation( { children } ) {
+	const [ currentView, setCurrentView ] = useState( VIEW_ROOT );
+
+	return (
+		<GlobalStylesNavigationContext.Provider
+			value={ { currentView, setCurrentView } }
```
Comment (sarayourfriend) [inline packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js:17]: Nit: should this object be memoized?

File: packages/edit-site/src/components/sidebar/global-styles-v2/views/colors.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export default function GlobalStylesViewColors() {
+	const { setCurrentView } = useGlobalStylesNavigationContext();
+
+	const navigateToColorsElement = () => setCurrentView( 'colors-element' );
+	const navigateToRoot = () => setCurrentView( 'root' );
+
+	return (
+		<>
+			<h3>Colors view</h3>
+			<button onClick={ navigateToRoot }>&lt; Back to Root view</button>
```
Comment (jorgefilipecosta) [inline packages/edit-site/src/components/sidebar/global-styles-v2/views/colors.js:15]: This is just for demo purposes for now, but it already shows that we probably need general components with things like the back button to go back to previews view, the title of the view etc.

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js (hunk 1 above)
Comment (jorgefilipecosta) [inline packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js:17]: ´{ currentView, setCurrentView }` creates a new object every time the component rerenders, given that this is a context value I guess it will trigger unrequited rerender on the components tree. Maybe in this case useMemo is helpful.

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+} from './constants';
+import { useGlobalStylesNavigationContext } from './context';
+import GlobalStylesViewColors from '../views/colors';
+import GlobalStylesViewColorsElement from '../views/colors-element';
+import GlobalStylesViewColorsPalette from '../views/colors-palette';
+import GlobalStylesViewRoot from '../views/root';
+import GlobalStylesViewTypography from '../views/typography';
+import GlobalStylesViewTypographyElement from '../views/typography-element';
+
+const allViews = {
```
Comment (jorgefilipecosta) [inline packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js:20]: This may grow huge with all the subviews. Maybe the root view does not need to be aware of the subviews and just of its children, and then its children are aware of its subviews.
Another alternative is importing the subviews from the children.
```
import { views as colorViews } from '../views/colors'

const allViews = {
	...colorViews,
}
```
But for now keeping them in a central place is ok we don't know yet if it is going to be a problem or not.

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	[ VIEW_ROOT ]: GlobalStylesViewRoot,
+	[ VIEW_COLORS ]: GlobalStylesViewColors,
+	[ VIEW_COLORS_ELEMENT ]: GlobalStylesViewColorsElement,
+	[ VIEW_COLORS_PALETTE ]: GlobalStylesViewColorsPalette,
+	[ VIEW_TYPOGRAPHY ]: GlobalStylesViewTypography,
```
Comment (jorgefilipecosta) [inline packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js:25]: I may want to change the typography settings of block A I just installed. So in practice all these views could also be nested e.g: I may go to root-> view typography to change typography settings for all the site, or I may go to root -> blocks -> block a (dynamic depending on the blocks of site) -> view typography.

It would be nice to see how this system expands to this dynamic case. E.g: we may have a `const all blocks = []` that simulates the blocks on the website. And then we could simulate this "dynamic" view case.
It is not a blocker we can git it a try in a follow-up PR.

**Review by sarayourfriend — APPROVED** [review]
This LGTM.

Just a couple notes from our sychronous discussion earlier:

We will manage a navigation stack of slugs (like view breadcrumbs) each time a View is entered into. Using these slugs we'll be able to retrieve any relevant metadata about the current and previous views. This will allow us to programatically set the view header and back buttons to the correct values as well as render the correct view when hitting the back buttons.

Do we also need to think of a `Link` type component like react-router has that will facilitate those navigations/pop things into the stack and update the current view? Or is using the `setCurrentView` function good enough?

**Review by jorgefilipecosta — COMMENTED** [review]
Thank you for proposing a simple navigation system @tyxla 👍 I really like this approach.

If possible it would be good to see how this system expands for the dynamic view/routes case as for sure we will need to have views per each block and it seems to only thing not covered.
We can simulate for example that inside the root we have a blocks view, and then inside this blocks view for each element of an array we have another view (for each block) and inside the view of each block, we can have all the global views again (excluding blocks for now because we don't have styles nesting yet).

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js (hunk 1 above)
Comment (tyxla) [reply packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js:17]: Good call 👍

File: packages/edit-site/src/components/sidebar/global-styles-v2/views/colors.js (hunk 2 above)
Comment (tyxla) [reply packages/edit-site/src/components/sidebar/global-styles-v2/views/colors.js:15]: Of course 👍

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js (hunk 1 above)
Comment (tyxla) [reply packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js:17]: Yep, that was an oversight of mine 😉

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js (hunk 3 above)
Comment (tyxla) [reply packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js:20]: Absolutely. Once we're done with this, it will have to be dynamic, otherwise it's just useless.

File: packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js (hunk 4 above)
Comment (tyxla) [reply packages/edit-site/src/components/sidebar/global-styles-v2/navigation/view.js:25]: Good call 👍  Once we have this running with the prototype, we can start experimenting with the nested and dynamic views, and it's a good idea to do it sooner than later.

**Comment by tyxla** [comment]
Lots of great feedback, thank you, folks! I think we have to take most (if not all) of your suggestions for the long-term navigation solution.

My point was to have an initial version of this so we can get the prototype running and to a place where we can actually see it, just like we can in the g2 repo.

My next steps are to merge this PR into #32923 and make them work together to allow spinning the new interface in a storybook. From there we can advance with component work, navigation, data, and all the other moving pieces in separate PRs.

**Comment by ciampo** [comment]
As a potential alternative to implementing a new simpler navigation, have we considered just keeping the routing + navigation system from the existing prototype?

Since the medium/long term plan is to come up with a more complex and comprehensive routing solution, for now we could just keep what is already in the prototype since it seems to be working well enough and has reasonable APIs  — and then swap the implementation with a more robust / well-thought-out at a later point

**Comment by tyxla** [comment]
> As a potential alternative to implementing a new simpler navigation, have we considered just keeping the routing + navigation system from the existing prototype?

I personally haven't, but I agree with giving it a try in any case. There's a lot we can learn from playing with moving it along from g2.

**Comment by ciampo** [comment]
> I personally haven't, but I agree with giving it a try in any case

Got it! The reason why I was suggesting to keep whatever routing solution is currently implemented in the prototype is that I wasn't aware that it is a blocker preventing the rest of the migration from happening

> One of the bigger blockers for moving #32923 forward (moving global styles v2 from g2 to Gutenberg) is the navigation/router logic

What aspects of the navigation/router logic are blocking the migration?

File: storybook/main.js [hunk 5]
```diff
@@ -8,6 +8,7 @@ const stories = [
 	'../packages/block-editor/src/**/stories/*.js',
 	'../packages/components/src/**/stories/*.js',
 	'../packages/icons/src/**/stories/*.js',
+	'../packages/edit-site/src/**/stories/*.js',
```
Comment (ciampo) [inline storybook/main.js:11]: Maybe I'm missing something, but why is this change needed in this PR?

File: packages/edit-site/src/components/sidebar/global-styles-v2/index.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+	return (
+		<GlobalStylesNavigation>
+			<GlobalStylesView />
+		</GlobalStylesNavigation>
+	);
+}
+
+function GlobalStylesView() {
+	return (
+		<Navigation data={ GLOBAL_STYLES_VIEWS } initial={ VIEW_ROOT }>
```
Comment (ciampo) [inline packages/edit-site/src/components/sidebar/global-styles-v2/index.js:23]: It seems like we're "defaulting" to `VIEW_ROOT` as the initial view in 2 places (here and when setting the initial value of `currentView` in `packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js`), should we aim at having it defined only in one place as the initial/starting view?

Comment (ciampo) [reply packages/edit-site/src/components/sidebar/global-styles-v2/index.js:23]: Also, I'm not very familiar with this component, but after a quick look I can't see the `data` and `initial` props being defined in `packages/components/src/navigation/index.js`

**Comment by tyxla** [comment]
> What aspects of the navigation/router logic are blocking the migration?

Just the fact that the components and functionality used for navigation and routing in g2 are not present in Gutenberg right now, and I'm not convinced we want to just move them here.

File: storybook/main.js (hunk 5 above)
Comment (tyxla) [reply storybook/main.js:11]: In order for the global styles story to be included, otherwise it's not.

File: packages/edit-site/src/components/sidebar/global-styles-v2/index.js (hunk 6 above)
Comment (tyxla) [reply packages/edit-site/src/components/sidebar/global-styles-v2/index.js:23]: > It seems like we're "defaulting" to `VIEW_ROOT` as the initial view in 2 places (here and when setting the initial value of `currentView` in `packages/edit-site/src/components/sidebar/global-styles-v2/navigation/index.js`), should we aim at having it defined only in one place as the initial/starting view?

I like that idea 👍 

> Also, I'm not very familiar with this component, but after a quick look I can't see the `data` and `initial` props being defined in `packages/components/src/navigation/index.js`

I think these are remnants of a previous approach that I didn't clean up. Can be removed.

**Review by jorgefilipecosta — APPROVED** [review]
Things look good to me 👍  This navigation system is simple to understand and does the job. I think we should try to test the case of the dynamic route as soon as possible as that seems to be the only remaining functionality we need from the router.

**Comment by tyxla** [comment]
Thank you all for the feedback!

As we discussed yesterday with @diegohaz, @sarayourfriend, @ciampo, and @griffbrad, I'd like to try out moving the g2 navigation components here early next week. Considering that their API is richer and close to `react-router`, we might have more options if we end up using them. So I'd like to see how difficult that would be and what it would take in order to weigh the different approaches.

**Comment by tyxla** [comment]
Now that we've used the navigation and routing from g2 in #32923, I think we can close this one.
